web/ui/react-app/src/PanelList.tsx

```diff
@@ -75,6 +79,40 @@ class PanelList extends Component<any, PanelListState> {
     }
   }
 
+  isHistoryEnabled = () => JSON.parse(localStorage.getItem('enable-query-history') || 'false') as boolean;
+
+  getHistoryItems = () => JSON.parse(localStorage.getItem('history') || '[]') as string[];
+
+  toggleQueryHistory = (enabled: boolean) => {
+    localStorage.setItem('enable-query-history', `${enabled}`);
+    this.setMetrics();
+  }
+
+  setMetrics = () => {
+    if (this.isHistoryEnabled()) {
+      const historyItems = this.getHistoryItems();
+      const { length } = historyItems;
+      this.setState({
+        metricNames: [...historyItems.slice(length - 50, length), ...this.initialMetricsNames],
```

Comment:
I think it'd be good to keep query history items separate from metric names somehow. We could either pass in two different properties (`metricNames` and `pastQueries`) or a single object in which autocompletion items have different types (`autocomplete={{metricNames: [...], pastQueries: [...]}}`). Because I'd like to differentiate them in the dropdown later. Super ugly mockup for demonstration purposes:

![differentiated_autocomplete](https://user-images.githubusercontent.com/538008/67292232-0af39580-f4e3-11e9-8b69-7d403174681e.png)

Reply:
Yeah, this looks pretty organized.

Reply:
@juliusv, I'm going to implement this in separate PR since too much things going on here :)

Reply:
@juliusv https://github.com/prometheus/prometheus/pull/6211 - splitting autocomplete dropdown into groups